visual-recognition/v3.ts

```diff
@@ -87,75 +86,84 @@ class VisualRecognitionV3 extends GeneratedVisualRecognitionV3 {
-    const _classifier_ids = _parameters.classifier_ids || params.classifier_ids;
-    const _owners = _parameters.owners || params.owners;
-    const _threshold = _parameters.threshold || params.threshold;
-    let _obj = {
-      url: _url,
-      classifier_ids: _classifier_ids,
-      owners: _owners,
-      threshold: _threshold
-    };
-    // remove null/undefined keys
-    Object.keys(_obj).forEach(key => {
-      _obj[key] == null && delete _obj[key];
+  }
+
+  private parseParameters(params) {
+    let parameters;
+    try {
+      parameters = JSON.parse(params.parameters);
+    } catch (e) {
+      parameters = {};
+    }
+    const parameters_keys = ['url', 'classifier_ids', 'owners', 'threshold'];
+    let _parameters = {};
+    parameters_keys.forEach(key => {
+      if (parameters[key] || params[key]) {
+        _parameters[key] = parameters[key] || params[key];
+      }
     });
-    return Object.keys(_obj).length > 0 ? _obj : null;
+    return Object.keys(_parameters).length > 0 ? _parameters : null;
   }
 
   classify(params, callback) {
-    const _parameters = this.parseParameters(params, callback) || {};
-    // set defaults for classify()
-    if (!_parameters.classifier_ids) {
-      _parameters.classifier_ids = ['default'];
+    const _callback = typeof callback === 'function' ? callback : () => {};
+    try {
+      this.checkVisionParams(params);
+    } catch (e) {
+      return _callback(e);
     }
-    if (!_parameters.owners) {
-      _parameters.owners = ['me', 'IBM'];
+    if (params.image_file) {
+      params.images_file = params.image_file;
     }
-    const _params = extend(params, { parameters: JSON.stringify(_parameters) });
-    return super.classify(_params, callback);
+    const defaultParameters = {
+      classifier_ids: ['default'],
+      owners: ['me', 'IBM']
+    };
+    const _parameters = extend(
+      {},
+      defaultParameters,
+      this.parseParameters(params)
+    );
+    const _params = extend(params, { parameters: _parameters });
+    return super.classify(_params, _callback);
   }
 
   detectFaces(params, callback) {
-    const _parameters = this.parseParameters(params, callback);
-    if (!_parameters) {
-      // if images_file
-      return super.detectFaces(params, callback);
-    } else {
-      const _params = extend(params, {
-        parameters: JSON.stringify(_parameters)
-      });
-      return super.detectFaces(_params, callback);
+    const _callback = typeof callback === 'function' ? callback : () => {};
+    try {
+      this.checkVisionParams(params);
+    } catch (e) {
+      return _callback(e);
+    }
+    if (params.image_file) {
+      params.images_file = params.image_file;
     }
+    const _params = extend({}, params, {
+      parameters: this.parseParameters(params)
+    });
+    return super.detectFaces(_params, _callback);
   }
 
   createClassifier(params, callback) {
     const _callback = callback ? callback : () => {};
+    const err = new Error(
```

Comment:
How do you handle errors due to missing parameters now?

Reply:
This is what we're doing in the services methods now.
```
{{#hasRequiredParams}}
    const _params = extend({}, params);
    const _callback = (callback) ? callback : () => {};
{{/hasRequiredParams}}
    {{^hasRequiredParams}}
    const _params = (typeof params === 'function' && !callback) ? {} : extend({}, params);
    const _callback = (typeof params === 'function' && !callback) ? params : (callback) ? callback : () => {};
    {{/hasRequiredParams}}
```